dpnp/backend/tests/test_broadcast_iterator.cpp

```diff
@@ -0,0 +1,202 @@
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include <gtest/gtest.h>
+#include <numeric>
+#include <vector>
+
+#include "dpnp_iterator.hpp"
+
+#define DPNP_LOCAL_QUEUE 1 // TODO need to fix build procedure and remove this workaround. Issue #551
+#include "queue_sycl.hpp"
+
+using namespace std;
+using dpnpc_it_t = DPNPC_id<size_t>::iterator;
+using dpnpc_value_t = dpnpc_it_t::value_type;
+using dpnpc_index_t = dpnpc_it_t::size_type;
+
+template <typename _DataType>
+vector<_DataType> get_input_data(const vector<dpnpc_index_t>& shape)
+{
+    const dpnpc_index_t size = accumulate(shape.begin(), shape.end(), dpnpc_index_t(1), multiplies<dpnpc_index_t>());
+
+    vector<_DataType> input_data(size);
+    iota(input_data.begin(), input_data.end(), 1); // let's start from 1 to avoid cleaned memory comparison
+
+    return input_data;
+}
+
+TEST(TestBroadcastIterator, take_value_broadcast_loop_2D)
+{
+    vector<dpnpc_value_t> expected_data{1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3};
+    // expected input data 1, 2, 3
+    vector<dpnpc_value_t> input_data = get_input_data<dpnpc_value_t>({3, 1});
+    DPNPC_id<dpnpc_value_t> result_obj(input_data.data(), {3, 1});
+    result_obj.broadcast_to_shape({3, 4});
+
+    for (dpnpc_index_t output_id = 0; output_id < result_obj.get_output_size(); ++output_id)
+    {
+        EXPECT_EQ(result_obj[output_id], expected_data[output_id]);
+    }
+}
+
+TEST(TestBroadcastIterator, take_value_broadcast_loop_3D)
+{
+    vector<dpnpc_value_t> expected_data{1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6};
+    // expected input data 1, 2, 3, 4, 5, 6
+    vector<dpnpc_value_t> input_data = get_input_data<dpnpc_value_t>({2, 3});
+    DPNPC_id<dpnpc_value_t> result_obj(input_data.data(), {2, 3});
+    result_obj.broadcast_to_shape({2, 2, 3});
+
+    for (dpnpc_index_t output_id = 0; output_id < result_obj.get_output_size(); ++output_id)
+    {
+        EXPECT_EQ(result_obj[output_id], expected_data[output_id]);
+    }
+}
+
+TEST(TestBroadcastIterator, output_size_broadcast)
+{
+    vector<dpnpc_value_t> input_data = get_input_data<dpnpc_value_t>({3, 4});
+    DPNPC_id<dpnpc_value_t> result_obj(input_data.data(), {3, 4});
+    result_obj.broadcast_to_shape({2, 3, 4});
+
+    const dpnpc_index_t output_size = result_obj.get_output_size();
+    EXPECT_EQ(output_size, 24);
+}
+
+struct IteratorParameters
+{
+    vector<dpnpc_it_t::size_type> input_shape;
+    vector<dpnpc_it_t::size_type> output_shape;
```

Comment:
Is it defined or standardized somewhere? I mean, `output_shape` as an input.
I see, at least, two approaches:
1. we keep input internally: `input_shape` and `output_shape`. Also we internally calculate step distances and other things to do iteration.
2. we keep input internally: `input_shape` and `broatcast_axis`. Also we internally calculate `output_shape`, step distances and other things to do iteration.

#2 will match existed interface code (to function `set_axes`) and, perhaps, required some extra parameter in ctor() like:
`DPNPC_id<dpnpc_value_t> result_obj(input_data.data(), {3, 4}, BROADCAST);`

Reply:
I don't know what complexity you are ready to implement and NumPy is required.
as far as I understand, we have two input shapes ("several" in general). Let's take:
```
shape1={2, 3, 4}
shape2={3, 4}
```
it looks like we can broadcast `shape2` into `shape1` by two steps
```
1. [3, 4] => [1, 3, 4]
2. [1, 3, 4] => [2, 3, 4]
```
Also, as I understand, shapes [3, 4] and [3, 1, 4, 1] have the same layout in memory.
I think everything depends on input parameters. If we have shapes as input - it is ok. I think having `axis` to broadcast is more flexible and scalable solution.

Reply:
From another perspective, it is NOT good idea.
May be better leave it as is but make function names more clear.